Flaky `DiskSpaceRecoveryIT$WithStandardContainerTest.shouldRecoverAfterOutOfDiskSpaceAfterExporting`

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? Increasingly often
- Does it block your work? Probably soon
- Do we suspect that it is a real failure? No

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
Expecting code not to raise a throwable but caught
  "io.camunda.zeebe.client.api.command.ClientStatusException: Cannot accept requests for partition 1. Broker is out of disk space
</pre>
</details>

**Hypotheses**

Caused by async segment preparation + offset log compaction which changes disk usage characteristics.

Looking at the logs, they make no sense. See for yourself:

```
18:22:05.485 [] [] [atomix-cluster-heartbeat-sender] DEBUG io.atomix.cluster.protocol.swim.probe - gateway - Requesting probe of Member{id=1, address=0.0.0.0:1894, properties={brokerInfo=EADJAAAABAABAAAAAwAAAAMAAAABAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoxODkzBQAADAAADgAAADguNS4wLVNOQVBTSE9UBQABAgAAAAA=}, version=8.5.0-SNAPSHOT, timestamp=1707330092446, state=SUSPECT, incarnationNumber=1707330092449} from [[Member{id=2, address=0.0.0.0:1898, properties={brokerInfo=EADJAAAABAACAAAAAwAAAAMAAAABAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoxODk3BQAADAAADgAAADguNS4wLVNOQVBTSE9UBQABAwAAAAA=}}, Member{id=0, address=0.0.0.0:1890, properties={brokerInfo=EADJAAAABAAAAAAAAwAAAAMAAAABAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoxODg5BQAADAAADgAAADguNS4wLVNOQVBTSE9UBQABAQAAAAA=}}]]
18:22:05.486 [] [] [netty-messaging-event-epoll-client-17] DEBUG io.atomix.cluster.protocol.swim.probe - gateway - Probe request of Member{id=1, address=0.0.0.0:1894, properties={brokerInfo=EADJAAAABAABAAAAAwAAAAMAAAABAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoxODkzBQAADAAADgAAADguNS4wLVNOQVBTSE9UBQABAgAAAAA=}, version=8.5.0-SNAPSHOT, timestamp=1707330092446, state=SUSPECT, incarnationNumber=1707330092449} from Member{id=0, address=0.0.0.0:1890, properties={brokerInfo=EADJAAAABAAAAAAAAwAAAAMAAAABAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoxODg5BQAADAAADgAAADguNS4wLVNOQVBTSE9UBQABAQAAAAA=}} failed
18:22:05.486 [] [] [netty-messaging-event-epoll-client-16] DEBUG io.atomix.cluster.protocol.swim.probe - gateway - Probe request of Member{id=1, address=0.0.0.0:1894, properties={brokerInfo=EADJAAAABAABAAAAAwAAAAMAAAABAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoxODkzBQAADAAADgAAADguNS4wLVNOQVBTSE9UBQABAgAAAAA=}, version=8.5.0-SNAPSHOT, timestamp=1707330092446, state=SUSPECT, incarnationNumber=1707330092449} from Member{id=2, address=0.0.0.0:1898, properties={brokerInfo=EADJAAAABAACAAAAAwAAAAMAAAABAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoxODk3BQAADAAADgAAADguNS4wLVNOQVBTSE9UBQABAwAAAAA=}} failed
18:22:05.486 [] [] [atomix-cluster-heartbeat-sender] INFO  io.atomix.cluster.protocol.swim.probe - gateway - Failed all probes of Member{id=1, address=0.0.0.0:1894, properties={brokerInfo=EADJAAAABAABAAAAAwAAAAMAAAABAAAAAAABCgAAAGNvbW1hbmRBcGkMAAAAMC4wLjAuMDoxODkzBQAADAAADgAAADguNS4wLVNOQVBTSE9UBQABAgAAAAA=}}. Marking as suspect.
18:22:06.486 [] [] [atomix-cluster-heartbeat-sender] WARN  io.atomix.cluster.protocol.swim.probe - gateway - Failed to probe 2
java.net.ConnectException: Failed to connect channel for address 0.0.0.0:1898 (resolved: null) : io.netty.channel.AbstractChannel$AnnotatedConnectException: finishConnect(..) failed: Connection refused: /0.0.0.0:1898
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$bootstrapClient$39(NettyMessagingService.java:753) ~[zeebe-atomix-cluster-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:590) ~[netty-common-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:583) ~[netty-common-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:559) ~[netty-common-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:492) ~[netty-common-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:636) ~[netty-common-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:629) ~[netty-common-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:118) ~[netty-common-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.fulfillConnectPromise(AbstractEpollChannel.java:675) ~[netty-transport-classes-epoll-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.finishConnect(AbstractEpollChannel.java:694) ~[netty-transport-classes-epoll-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.epollOutReady(AbstractEpollChannel.java:567) ~[netty-transport-classes-epoll-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:499) ~[netty-transport-classes-epoll-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:407) ~[netty-transport-classes-epoll-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.106.Final.jar:4.1.106.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.106.Final.jar:4.1.106.Final]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
```

Why are we seeing the gateway trying to probe a whole cluster? Member 1? Member 2? This test only starts a single container with embedded gateway and no other containers...

It seems there might be something broken with the logging 🤔 

Thankfully, the job logs contain the actual error:

```
2024-02-07T18:22:22.7242630Z [ERROR] io.camunda.zeebe.it.health.DiskSpaceRecoveryIT$WithAlreadyDiskFullTest.shouldNotProcessWhenOutOfDiskSpaceOnStart -- Time elapsed: 17.77 s <<< FAILURE!
2024-02-07T18:22:22.7244426Z org.opentest4j.AssertionFailedError: 
2024-02-07T18:22:22.7244893Z 
2024-02-07T18:22:22.7245148Z Expecting a root cause with message:
2024-02-07T18:22:22.7246323Z   "RESOURCE_EXHAUSTED: Cannot accept requests for partition 1. Broker is out of disk space"
2024-02-07T18:22:22.7247363Z but message was:
2024-02-07T18:22:22.7249051Z   "DEADLINE_EXCEEDED: deadline exceeded after 9.999644600s. Name resolution delay 0.000109250 seconds. [closed=[], open=[[buffered_nanos=3067120, remote_addr=localhost/127.0.0.1:32941]]]".
2024-02-07T18:22:22.7250743Z Throwable that failed the check:
2024-02-07T18:22:22.7251202Z 
2024-02-07T18:22:22.7252824Z io.camunda.zeebe.client.api.command.ClientStatusException: deadline exceeded after 9.999644600s. Name resolution delay 0.000109250 seconds. [closed=[], open=[[buffered_nanos=3067120, remote_addr=localhost/127.0.0.1:32941]]]
2024-02-07T18:22:22.7255810Z 	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.transformExecutionException(ZeebeClientFutureImpl.java:116)
2024-02-07T18:22:22.7257697Z 	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:54)
2024-02-07T18:22:22.7275537Z 	at io.camunda.zeebe.it.health.DiskSpaceRecoveryIT.publishMessage(DiskSpaceRecoveryIT.java:74)
2024-02-07T18:22:22.7276634Z 	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:63)
2024-02-07T18:22:22.7277539Z 	at org.assertj.core.api.AssertionsForClassTypes.catchThrowable(AssertionsForClassTypes.java:892)
2024-02-07T18:22:22.7278486Z 	at org.assertj.core.api.Assertions.catchThrowable(Assertions.java:1365)
2024-02-07T18:22:22.7279287Z 	at org.assertj.core.api.Assertions.assertThatThrownBy(Assertions.java:1209)
2024-02-07T18:22:22.7280407Z 	at io.camunda.zeebe.it.health.DiskSpaceRecoveryIT$WithAlreadyDiskFullTest.shouldNotProcessWhenOutOfDiskSpaceOnStart(DiskSpaceRecoveryIT.java:146)
2024-02-07T18:22:22.7281672Z 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
2024-02-07T18:22:22.7282509Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
2024-02-07T18:22:22.7283251Z 	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:728)
2024-02-07T18:22:22.7284291Z 	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
2024-02-07T18:22:22.7285393Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
2024-02-07T18:22:22.7286502Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
2024-02-07T18:22:22.7287614Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
2024-02-07T18:22:22.7288642Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
2024-02-07T18:22:22.7289992Z 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
2024-02-07T18:22:22.7292248Z 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
2024-02-07T18:22:22.7294070Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
2024-02-07T18:22:22.7295467Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
2024-02-07T18:22:22.7296734Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
2024-02-07T18:22:22.7298690Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
2024-02-07T18:22:22.7300563Z 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
2024-02-07T18:22:22.7301916Z 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
2024-02-07T18:22:22.7303363Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:218)
2024-02-07T18:22:22.7305533Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2024-02-07T18:22:22.7307576Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:214)
2024-02-07T18:22:22.7309567Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:139)
2024-02-07T18:22:22.7311648Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:69)
2024-02-07T18:22:22.7313606Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
2024-02-07T18:22:22.7315803Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2024-02-07T18:22:22.7317730Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
2024-02-07T18:22:22.7319352Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
2024-02-07T18:22:22.7320991Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
2024-02-07T18:22:22.7322901Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2024-02-07T18:22:22.7324893Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
2024-02-07T18:22:22.7326700Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
2024-02-07T18:22:22.7329101Z 	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:202)
2024-02-07T18:22:22.7332523Z 	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.invokeAll(ForkJoinPoolHierarchicalTestExecutorService.java:146)
2024-02-07T18:22:22.7335050Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
2024-02-07T18:22:22.7336817Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2024-02-07T18:22:22.7337903Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
2024-02-07T18:22:22.7339412Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
2024-02-07T18:22:22.7341172Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
2024-02-07T18:22:22.7343578Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2024-02-07T18:22:22.7345767Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
2024-02-07T18:22:22.7347901Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
2024-02-07T18:22:22.7350564Z 	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:202)
2024-02-07T18:22:22.7354125Z 	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.executeNonConcurrentTasks(ForkJoinPoolHierarchicalTestExecutorService.java:172)
2024-02-07T18:22:22.7357331Z 	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.invokeAll(ForkJoinPoolHierarchicalTestExecutorService.java:152)
2024-02-07T18:22:22.7358732Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
2024-02-07T18:22:22.7359859Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2024-02-07T18:22:22.7360983Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
2024-02-07T18:22:22.7361916Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
2024-02-07T18:22:22.7362842Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
2024-02-07T18:22:22.7364186Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2024-02-07T18:22:22.7366087Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
2024-02-07T18:22:22.7367278Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
2024-02-07T18:22:22.7369679Z 	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:202)
2024-02-07T18:22:22.7372390Z 	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.invokeAll(ForkJoinPoolHierarchicalTestExecutorService.java:146)
2024-02-07T18:22:22.7374154Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
2024-02-07T18:22:22.7375872Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2024-02-07T18:22:22.7377012Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
2024-02-07T18:22:22.7377977Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
2024-02-07T18:22:22.7378921Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
2024-02-07T18:22:22.7380018Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2024-02-07T18:22:22.7381105Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
2024-02-07T18:22:22.7382109Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
2024-02-07T18:22:22.7383442Z 	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:202)
2024-02-07T18:22:22.7384839Z 	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
2024-02-07T18:22:22.7385583Z 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
2024-02-07T18:22:22.7386612Z 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
2024-02-07T18:22:22.7387422Z 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
2024-02-07T18:22:22.7388160Z 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
2024-02-07T18:22:22.7389207Z 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
2024-02-07T18:22:22.7390778Z Caused by: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 9.999644600s. Name resolution delay 0.000109250 seconds. [closed=[], open=[[buffered_nanos=3067120, remote_addr=localhost/127.0.0.1:32941]]]
2024-02-07T18:22:22.7392459Z 	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
2024-02-07T18:22:22.7393300Z 	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
2024-02-07T18:22:22.7394276Z 	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:52)
2024-02-07T18:22:22.7394864Z 	... 73 more
2024-02-07T18:22:22.7395894Z Caused by: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 9.999644600s. Name resolution delay 0.000109250 seconds. [closed=[], open=[[buffered_nanos=3067120, remote_addr=localhost/127.0.0.1:32941]]]
2024-02-07T18:22:22.7397050Z 	at io.grpc.Status.asRuntimeException(Status.java:533)
2024-02-07T18:22:22.7397765Z 	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:481)
2024-02-07T18:22:22.7398619Z 	at io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:489)
2024-02-07T18:22:22.7399490Z 	at io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:453)
2024-02-07T18:22:22.7400368Z 	at io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:486)
2024-02-07T18:22:22.7401186Z 	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:574)
2024-02-07T18:22:22.7401884Z 	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:72)
2024-02-07T18:22:22.7402820Z 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:742)
2024-02-07T18:22:22.7404205Z 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
2024-02-07T18:22:22.7405077Z 	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
2024-02-07T18:22:22.7405855Z 	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
2024-02-07T18:22:22.7406680Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
2024-02-07T18:22:22.7407712Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
2024-02-07T18:22:22.7408410Z 	at java.base/java.lang.Thread.run(Thread.java:1583)
2024-02-07T18:22:22.7408713Z 
2024-02-07T18:22:22.7409287Z 	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
2024-02-07T18:22:22.7410296Z 	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
2024-02-07T18:22:22.7411492Z 	at io.camunda.zeebe.it.health.DiskSpaceRecoveryIT$WithAlreadyDiskFullTest.shouldNotProcessWhenOutOfDiskSpaceOnStart(DiskSpaceRecoveryIT.java:147)
2024-02-07T18:22:22.7412505Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
2024-02-07T18:22:22.7413196Z 	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
2024-02-07T18:22:22.7414442Z 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
2024-02-07T18:22:22.7415849Z 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
2024-02-07T18:22:22.7416878Z 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
2024-02-07T18:22:22.7417691Z 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
2024-02-07T18:22:22.7419118Z 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

We can see that sometimes the command times out instead of failing with the expected error. I think this can happen if the node is overloaded (we run DinD after all), so the test may be slower.

So I would propose two things, since here we're not testing for performance:

1. Increase the request timeout
2. Retry deadline exceeded up to 3 times.

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/health/DiskSpaceRecoveryIT.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/health/DiskSpaceRecoveryIT.java
@@ -27,6 +27,7 @@ import java.util.Map;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 import org.agrona.CloseHelper;
+import org.awaitility.Awaitility;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Nested;
@@ -80,7 +81,6 @@ final class DiskSpaceRecoveryIT {
         .correlationKey(String.valueOf(1))
         .variables(Map.of("key", "abc".repeat(4096)))
         .timeToLive(Duration.ZERO)
-        .requestTimeout(Duration.ofSeconds(30))
         .send()
         .join();
   }

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/health/DiskSpaceRecoveryIT.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/health/DiskSpaceRecoveryIT.java
@@ -7,14 +7,18 @@
  */
 package io.camunda.zeebe.it.health;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatNoException;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
 
 import io.camunda.zeebe.client.ZeebeClient;
+import io.camunda.zeebe.client.api.command.ClientStatusException;
 import io.camunda.zeebe.qa.util.actuator.PartitionsActuator;
 import io.camunda.zeebe.qa.util.testcontainers.ZeebeTestContainerDefaults;
 import io.camunda.zeebe.test.util.socket.SocketUtil;
+import io.camunda.zeebe.test.util.testcontainers.ContainerLogsDumper;
+import io.grpc.Status.Code;
 import io.zeebe.containers.ZeebeContainer;
 import io.zeebe.containers.ZeebeVolume;
 import io.zeebe.containers.engine.ContainerEngine;
@@ -27,6 +31,7 @@ import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
 import org.testcontainers.junit.jupiter.Container;
 import org.testcontainers.junit.jupiter.Testcontainers;
 
@@ -51,6 +56,11 @@ final class DiskSpaceRecoveryIT {
           .withEnv("ZEEBE_BROKER_DATA_DISK_FREESPACE_PROCESSING", "8MB")
           .withEnv("ZEEBE_BROKER_DATA_DISK_FREESPACE_REPLICATION", "1MB");
 
+  @SuppressWarnings("JUnitMalformedDeclaration")
+  @RegisterExtension
+  private final ContainerLogsDumper logsDumper =
+      new ContainerLogsDumper(() -> Map.of("broker", container));
+
   private ZeebeClient client;
 
   @AfterEach
@@ -70,6 +80,7 @@ final class DiskSpaceRecoveryIT {
         .correlationKey(String.valueOf(1))
         .variables(Map.of("key", "abc".repeat(4096)))
         .timeToLive(Duration.ZERO)
+        .requestTimeout(Duration.ofSeconds(30))
         .send()
         .join();
   }
@@ -142,10 +153,32 @@ final class DiskSpaceRecoveryIT {
 
     @Test
     void shouldNotProcessWhenOutOfDiskSpaceOnStart() {
-      // when - then
-      assertThatThrownBy(DiskSpaceRecoveryIT.this::publishMessage)
-          .hasRootCauseMessage(
-              "RESOURCE_EXHAUSTED: Cannot accept requests for partition 1. Broker is out of disk space");
+      // given
+      var retryCount = 0;
+
+      // when
+      while (retryCount < 3) {
+        try {
+          publishMessage();
+        } catch (final ClientStatusException e) {
+          retryCount++;
+
+          if (e.getStatusCode() == Code.DEADLINE_EXCEEDED) {
+            continue;
+          }
+
+          // then
+          assertThat(e.getStatusCode()).isEqualTo(Code.RESOURCE_EXHAUSTED);
+          assertThat(e)
+              .hasRootCauseMessage(
+                  "RESOURCE_EXHAUSTED: Cannot accept requests for partition 1. Broker is out of disk space");
+        }
+      }
+
+      assertThat(retryCount)
+          .as(
+              "Expected at least one out of three requests to not timeout; the container may be broken")
+          .isLessThan(4);
     }
   }
 }

